### ISSUE TITLE
Is there any support plan for yolov10?

### ISSUE THREAD

**Issue by tms2003** [marcoslucianops/DeepStream-Yolo#556]
Is there any support plan for yolov10?It looks like it should be faster

**Comment by reno77** [comment]
There's a forked repo here with yolov10 support , https://github.com/shashikant-ghangare/DeepStream-Yolo/tree/add-yolov10

**Comment by marcoslucianops** [comment]
I will the able to add the support 6 months after the launch date (23 May 2024).

**Comment by marcoslucianops** [comment]
I will the able to add the support in the next month. I can't post the new models when they are released because of some contracts I have.

**Comment by RadouaneK** [comment]
@marcoslucianops I hope you're still planning to support yolov10

**Comment by YangHR0805** [comment]
> There's a forked repo here with yolov10 support , https://github.com/shashikant-ghangare/DeepStream-Yolo/tree/add-yolov10
> 
> Edit: I tried it, the .onnx and .engine files got generated but I'm not getting any detections for my videos. Not sure why yet.


`import os
import onnx
import torch
import torch.nn as nn

from copy import deepcopy


class DeepStreamOutput(nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x):
        x = x['one2one'][0]
        x = x.transpose(1, 2)
        boxes = x[:, :, :4]
        convert_matrix = torch.tensor(
            [[1, 0, 1, 0], [0, 1, 0, 1], [-0.5, 0, 0.5, 0], [0, -0.5, 0, 0.5]], dtype=boxes.dtype, device=boxes.device
        )
        boxes @= convert_matrix
        print(f"boxes shape ------------------ is {boxes.shape}")
        scores, labels = torch.max(x[:, :, 4:], dim=-1, keepdim=True)
        return torch.cat([boxes, scores, labels.to(boxes.dtype)], dim=-1)


def yolov10_export(weights, device, inplace=True, fuse=True):
    ckpt = torch.load(weights, map_location='cpu')
    ckpt = (ckpt.get('ema') or ckpt['model']).to(device).float()
    if not hasattr(ckpt, 'stride'):
        ckpt.stride = torch.tensor([32.])
    if hasattr(ckpt, 'names') and isinstance(ckpt.names, (list, tuple)):
        ckpt.names = dict(enumerate(ckpt.names))
    model = ckpt.fuse().eval() if fuse and hasattr(ckpt, 'fuse') else ckpt.eval()
    for m in model.modules():
        t = type(m)
        if hasattr(m, 'inplace'):
            m.inplace = inplace
        elif t.__name__ == 'Upsample' and not hasattr(m, 'recompute_scale_factor'):
            m.recompute_scale_factor = None
        model = deepcopy(model).to(device)
        for p in model.parameters():
            p.requires_grad = False
        model.eval()
        model.float()
        model = model.fuse()
        for k, m in model.named_modules():
            if m.__class__.__name__ in ('Detect', 'RTDETRDecoder'):
                m.dynamic = True
                m.export = True
                m.format = 'onnx'
            elif m.__class__.__name__ == 'C2f':
                m.forward = m.forward_split
    return model


def suppress_warnings():
    import warnings
    warnings.filterwarnings('ignore', category=torch.jit.TracerWarning)
    warnings.filterwarnings('ignore', category=UserWarning)
    warnings.filterwarnings('ignore', category=DeprecationWarning)
    warnings.filterwarnings('ignore', category=FutureWarning)
    warnings.filterwarnings('ignore', category=ResourceWarning)


def main(args):
    suppress_warnings()

    input_file = args.weights
    onnx_output_file = os.path.join(os.path.dirname(input_file), 'ds_' + os.path.basename(input_file).replace('.pt', '.onnx'))
    labels_file = os.path.join(os.path.dirname(input_file), 'labels.txt')
    
    print(f'input_file is {input_file}')
    print(f'output_file is {onnx_output_file}')


    print(f'\nStarting: {input_file}')

    print('Opening YOLOv5 model')

    device = torch.device('cpu')
    model = yolov10_export(input_file, device)

    if len(model.names.keys()) > 0:
        print('Creating labels.txt file')
        with open(labels_file, 'w', encoding='utf-8') as f:
            for name in model.names.values():
                f.write(f'{name}\n')

    model = nn.Sequential(model, DeepStreamOutput())

    img_size = args.size * 2 if len(args.size) == 1 else args.size

    if img_size == [640, 640] and args.p6:
        img_size = [1280] * 2

    onnx_input_im = torch.zeros(16, 3, *img_size).to(device)

    print(f'onnx_input_im shape is {onnx_input_im.shape}')
    model.eval()

    output = model(onnx_input_im)
    print(f'output shape is {output[0].shape}')


    dynamic_axes = {
        'input': {
            0: 'batch'
        },
        'output': {
            0: 'batch'
        }
    }

    print('Exporting the model to ONNX')
    # torch.onnx.export(
    #     model, onnx_input_im, onnx_output_file, verbose=False, opset_version=args.opset, do_constant_folding=True,
    #     input_names=['input'], output_names=['output'], dynamic_axes=dynamic_axes if args.dynamic else None
    # )
    traced_model = torch.jit.trace(model, onnx_input_im)
    torch.onnx.export(traced_model, onnx_input_im, onnx_output_file, verbose=False, opset_version=12,
                          training=torch.onnx.TrainingMode.EVAL,
                          do_constant_folding=True,
                          input_names=['images'],
                          output_names=['output'],
                          dynamic_axes={'images': {0: 'batch', 2: 'height', 3: 'width'},  # shape(1,3,640,640)
                                        'output': {0: 'batch', 1: 'anchors'}  # shape(1,25200,85)
                                        } if args.dynamic else None)




    if args.simplify:
        print('Simplifying the ONNX model')
        import onnxslim
        model_onnx = onnx.load(onnx_output_file)
        model_onnx = onnxslim.slim(model_onnx)
        onnx.save(model_onnx, onnx_output_file)

    print(f'Done: {onnx_output_file}\n')
    return onnx_output_file


def parse_args():
    import argparse
    parser = argparse.ArgumentParser(description='DeepStream YOLOv5 conversion')
    parser.add_argument('-w', '--weights', required=True, type=str, help='Input weights (.pt) file path (required)')
    parser.add_argument('-s', '--size', nargs='+', type=int, default=[640], help='Inference size [H,W] (default [640])')
    parser.add_argument('--p6', action='store_true', help='P6 model')
    parser.add_argument('--opset', type=int, default=17, help='ONNX opset version')
    parser.add_argument('--simplify', action='store_true', help='ONNX simplify model')
    parser.add_argument('--dynamic', action='store_true', help='Dynamic batch-size')
    parser.add_argument('--batch', type=int, default=1, help='Static batch-size')
    args = parser.parse_args()
    if not os.path.isfile(args.weights):
        raise SystemExit('Invalid weights file')
    if args.dynamic and args.batch > 1:
        raise SystemExit('Cannot set dynamic batch-size and static batch-size at same time')
    return args


def infer(onnx_file, args):
    import onnxruntime as ort
    import numpy as np
    session = ort.InferenceSession(onnx_file)
    input_name = session.get_inputs()[0].name
    output_name = session.get_outputs()[0].name
    # input_shape = session.get_inputs()[0].shape
    # print(f'input_shape is {input_shape}')
    input_data = torch.zeros(8, 3, *args.size).numpy()
    output = session.run([output_name], {input_name: input_data})
    print(f'output len is {len(output)}')
    print(f'output shape is {output[0].shape}')

if __name__ == '__main__':
    args = parse_args()
    onnx_output_file = main(args)`

try this, it works for me.
